windows/gce-gpu-monitoring-cuda.ps1

```diff
@@ -0,0 +1,227 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https=//www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+ 
+<#
+This script can be used for monitoring  GPU utilization on Google Compute Engine
+instances with GPUs attached. It requires the nvidia-smi tool to be properly
+installed on the system (https=//developer.nvidia.com/nvidia-system-management-interface).
+
+This script assumes it is being executed on GCE instance and that it has access
+to the metadata server (https=//cloud.google.com/compute/docs/storing-retrieving-metadata).
+#>
+ 
+ Param(
+    #configure the interval, default is 5 and configure bigger than 5 to avoid throttling
+    $interval = 10,
+
+    #cuda is not supported yet
+    [switch]$collect_each_core = $false
+)
+
+$nvidia_smi_path = "C:\Program Files\NVIDIA Corporation\NVSMI\nvidia-smi.exe"
+$metadata_server = "http://metadata/computeMetadata/v1/instance/"
+$access_token_command = 'gcloud auth application-default print-access-token'
+
+$gpu_metrics = @{
+    'utilization.gpu'= 'instance/gpu/utilization';
+    'utilization.memory'= 'instance/gpu/memory_utilization';
+    'memory.total'= 'instance/gpu/memory_total';
+    'memory.used'= 'instance/gpu/memory_used';
+    'memory.free'= 'instance/gpu/memory_free';
+    'temperature.gpu'= 'instance/gpu/temperature';
+}
+
+
+function get_metrics{
+    <#
+    Calls the nvidia-smi tool to retrieve usage metrics about the
+    attached GPUs.
+    :return:
+    Dictionary that maps (gpu_type, gpu_bus_id) to metric values.
+    #>
+    
+    $query_params = 'gpu_name,gpu_bus_id,'+(($gpu_metrics.Keys|sort-object) -join ",")
+
+    $result = & $nvidia_smi_path --query-gpu=$query_params --format=csv,nounits,noheader
```

Comment:
If the format is csv, might I suggest us looking into using [`ConvertFrom-CSV` cmdlet](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-csv) to turn the result into an object that's easier to work with?

Reply:
Actually, I tried it first, but, it didn't fit to my usage.
The reason follows
1. the header info was not the same with what I have to use. I had to be on par with the column naming of Linux version and the column names of the command don't match to the one of Linux version.
2. The structure I want to form from the CSV isn't actually flat. For example,  the CSV output follows
ColumnA ColumnB ColumnC ColumnD ColumnE
firstname lastname 4               0              100
The structure I want to process later is
@{"firstname,lastname"= @{"columnC"= "4","ColumnD"="0","ColumnE"="100"}}